presto-raptorx/src/main/java/com/facebook/presto/raptorx/transaction/Transaction.java

```diff
@@ -0,0 +1,551 @@
+    private final SetMultimap<Long, String> viewsRemoved = HashMultimap.create();
+
+    private final Map<Long, Map<Long, ChunkInfo>> chunksAdded = new HashMap<>();
+    private final Map<Long, Set<Long>> chunksRemoved = new HashMap<>();
+
+    public Transaction(Metadata metadata, long transactionId, long commitId)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.transactionId = transactionId;
+        this.commitId = commitId;
+    }
+
+    public long getTransactionId()
+    {
+        return transactionId;
+    }
+
+    public long getCommitId()
+    {
+        return commitId;
+    }
+
+    public OptionalLong getRegisteredTransactionId()
+    {
+        return registeredTableIds.isEmpty() ? OptionalLong.empty() : OptionalLong.of(transactionId);
+    }
+
+    public void registerTable(long tableId)
+    {
+        if (registeredTableIds.isEmpty()) {
+            metadata.registerTransaction(transactionId);
+        }
+        if (registeredTableIds.add(tableId)) {
+            metadata.registerTransactionTable(transactionId, tableId);
+        }
+    }
+
+    public List<Action> getActions()
+    {
+        return ImmutableList.copyOf(actions);
+    }
+
+    public long nextTableId()
+    {
+        return metadata.nextTableId();
+    }
+
+    public long nextColumnId()
+    {
+        return metadata.nextColumnId();
+    }
+
+    public Optional<Long> getSchemaId(String schemaName)
+    {
+        return schemaIds.computeIfAbsent(schemaName, x -> metadata.getSchemaId(commitId, schemaName));
+    }
+
+    public Optional<Long> getTableId(SchemaTableName table)
+    {
+        return getSchemaId(table.getSchemaName())
+                .flatMap(schemaId -> getTableId(schemaId, table.getTableName()));
+    }
+
+    private Optional<Long> getTableId(long schemaId, String tableName)
+    {
+        return tableIds.computeIfAbsent(schemaId, x -> new HashMap<>())
+                .computeIfAbsent(tableName, x -> metadata.getTableId(commitId, schemaId, tableName));
+    }
+
+    public Optional<Long> getViewId(SchemaTableName view)
+    {
+        return getSchemaId(view.getSchemaName())
+                .flatMap(schemaId -> getViewId(schemaId, view.getTableName()));
+    }
+
+    private Optional<Long> getViewId(long schemaId, String viewName)
+    {
+        return viewIds.computeIfAbsent(schemaId, x -> new HashMap<>())
+                .computeIfAbsent(viewName, x -> metadata.getViewId(commitId, schemaId, viewName));
+    }
+
+    public List<String> listSchemas()
+    {
+        return concat(
+                metadata.listSchemas(commitId).stream(),
+                schemasAdded.stream())
+                .filter(not(schemasRemoved::contains))
+                .sorted()
+                .collect(toList());
+    }
+
+    public List<SchemaTableName> listTables()
+    {
+        // TODO: this should fetch all tables at once
+        return listSchemas().stream()
```

Comment:
"highker on Jan 30  Member
Add new dao method

use information_schema;
select * from tables;"

This is indeed valid, as the "TODO" speak the same. Right now it's not isolated, not actually RepeatableRead for all schemas. 

But, let's open another task to do the general topic "complete RepeatableRead". For now, all critical stuff tables chunks are correct;  for others like schema, use the task to investigate them all.